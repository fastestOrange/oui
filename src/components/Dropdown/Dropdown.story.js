import React from 'react';
import styled from 'styled-components';

import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, number } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';

import Dropdown from './index.js';
import BlockList from '../BlockList';
import TextField from '../TextField';
import Icon from 'react-oui-icons';

const data = [
  {title: 'Manual', description: 'Dolcelatte cheeseburger swiss paneer cow gouda edam cheese slices'},
  {title: 'Maximize Conventions', header: 'Automation',
    description: 'Everyone loves cut the cheese cow squirty cheese emmental jarlsberg halloumi camembert de normandie.'},
  {title: 'Faster Results', description: 'Cauliflower cheese stilton paneer hard cheese fondue.'},
];

const stories = storiesOf('Dropdown', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories.add('text behind', withInfo()(() => {
  return (
    <Container>
      <SubContainer>
        <ScrollContainer>
          <Dropdown
            arrowIcon={ true }
            isDisabled={ boolean('isDisabled', false) }
            buttonContent='Dropdown'
            width={ number('width', 300) }>
            <BlockList>
              {
                data.map((item, index) => {
                  return (
                    <BlockList.Category header={ item.header } key={ index }>
                      <BlockList.Item onClick={ action('click on complex item') }>
                        <div className="flex flex-align--center">
                          <div className="flex--1">
                            <div>{ item.title }</div>
                            <div className="muted micro">{ item.description }</div>
                          </div>
                        </div>
                      </BlockList.Item>
                    </BlockList.Category>
                  );
                })
              }
            </BlockList>
          </Dropdown>
          <h1>This text should be behind the open dropdown</h1>
        </ScrollContainer>
      </SubContainer>
    </Container>
  );
}));


stories.add('with icon', withInfo()(() => {
  return (
    <Container>
      <SubContainer>
        <ScrollContainer>
          <div style={{ width: 200, float: 'left' }}>
            <TextField
              placeholder=".tags"
              type="text"
            />
          </div>
          <Dropdown
            isDisabled={ boolean('isDisabled', false) }
            fullWidth={ boolean('fullWidth', false) }
            buttonContent={ <Icon name='hamburger' /> }
            width={ number('width', 350) }>
            <BlockList>
              {
                data.map((item, index) => {
                  return (
                    <BlockList.Category header={ item.header } key={ index }>
                      <BlockList.Item onClick={ action('click on complex item') }>
                        <div className="flex flex-align--center">
                          <div className="flex--1">
                            <div>{ item.title }</div>
                            <div className="muted micro">{ item.description }</div>
                          </div>
                        </div>
                      </BlockList.Item>
                    </BlockList.Category>
                  );
                })
              }
            </BlockList>
          </Dropdown>
          <h1>This text should be behind the open dropdown</h1>
        </ScrollContainer>
      </SubContainer>
    </Container>
  );
}));

const Container = styled.div`
  display: flex;
  flex: 1;
  height: 100vh;
  background: #f6f6f6;
`;

const SubContainer = styled.div`
  width: 700px;
  height: 400px;
  background: lightgrey;
  margin: 100px auto;
  overflow: scroll;
  padding: 20px;
  padding-top: 300px;
`;

const ScrollContainer = styled.div`
  width: 1000px;
  height: 1000px;
`;
