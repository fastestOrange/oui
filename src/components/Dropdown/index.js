import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Manager, Popper, Target } from 'react-popper';
import { withToggle } from '../../utils/recompose-utils';

const Dropdown = withToggle(({
  arrowIcon,
  buttonContent,
  children,
  hide,
  isDisabled = false,
  isOpen,
  fullWidth,
  placement = 'bottom-start',
  show,
  style,
  testSection,
  width = 200,
  zIndex = 999,
  toggle,
}) => {

  const groupClass = classNames(
    'oui-dropdown-group',
    { ['width--1-1']: fullWidth }
  );

  const buttonClass = classNames(
    'oui-button',
    { [`oui-button--${style}`]: style },
    { ['oui-button--full soft--left text--left']: fullWidth }
  );

  const iconClass = classNames(
    'push-half--left',
    {
      ['oui-arrow-inline--down']: arrowIcon === true || arrowIcon === 'down',
      ['oui-arrow-inline--left']: arrowIcon === 'left',
      ['oui-arrow-inline--right']: arrowIcon === 'right',
      ['oui-arrow-inline--up']: arrowIcon === 'up',
    }
  );

  return (
    <Manager
      data-ui-component={ true }
      className={ groupClass }
      data-test-section={ testSection }>
      <Target>
        <button
          type='button'
          className={ buttonClass }
          disabled={ isDisabled }
          onClick={ toggle }
          onBlur={ hide }>
          <div className='flex'>
            <div className='flex--1 truncate'>{ buttonContent }</div>
            {
              !!arrowIcon && arrowIcon !== 'none' && (
                <div className='text--right'><span className={ iconClass }></span></div>
              )
            }
          </div>
        </button>
      </Target>
      <Popper
        placement={ placement }
        className='oui-dropdown-children'
        style={{
          zIndex: zIndex,
          position: 'absolute',
          width: width,
          marginTop: 2,
          marginBottom: 2,
          boxShadow: '0 2px 3px rgba(0,0,0,.1)',
        }}
        onClick={ hide } >
        { isOpen && !isDisabled && children }
      </Popper>
    </Manager>
  );
});

Dropdown.propTypes = {
  /** Arrow icon direction:
    * - Defaults to 'none', which hides the arrow
    * - passing a prop value of false also hides the arrow
    * - passing a prop value of true uses down arrow
    */
  arrowIcon: PropTypes.oneOf([
    'down',
    'left',
    'none',
    'right',
    'up',
  ]),
  /** Button text, can be a string or element. */
  buttonContent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  /** Dropdown contents, typically using the <Blocklist> component. */
  children: PropTypes.node.isRequired,
  /** Button width is either full or inline-block. */
  fullWidth: PropTypes.bool,
  /** Unused... */
  handleClick: PropTypes.func,
  /** Disable button. */
  isDisabled: PropTypes.bool,
  /** Popper placement property */
  placement: PropTypes.oneOf([
    'top',
    'top-start',
    'top-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'right',
    'right-start',
    'right-end',
    'left',
    'left-start',
    'left-end',
  ]),
  /** Button style, e.g. highlight, danger, outline. */
  style: PropTypes.string,
  /** For automated testing only. */
  testSection: PropTypes.string,
  /** Dropdown menu width, in pixels. */
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /** Override default dropdown menu z-index. */
  zIndex: PropTypes.number,
};

Dropdown.defaultProps = {
  arrowIcon: 'none',
};

export default Dropdown;
