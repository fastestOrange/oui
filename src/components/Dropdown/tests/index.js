import React from 'react';
import Dropdown from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

const data = [
  {title: 'Manual', description: 'Dolcelatte cheeseburger swiss paneer cow gouda edam cheese slices'},
  {title: 'Maximize Conventions', header: 'Automation',
    description: 'Everyone loves cut the cheese cow squirty cheese emmental jarlsberg halloumi camembert de normandie.'},
  {title: 'Faster Results', description: 'Cauliflower cheese stilton paneer hard cheese fondue.'},
];

describe('components/Dropdown', () => {
  it('should render children when isOpen is true', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should not render children when isDisabled is true', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        idDisabled={ true }>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should not use .oui-arrow-inline--down when icon isEqual to triangle', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should use oui-button--full class when fullWidth is true', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        fullWidth={ true }>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should highlight button class when style equals highlight', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        style='highlight'>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});

