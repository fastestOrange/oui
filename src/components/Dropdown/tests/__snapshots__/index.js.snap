// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Dropdown should highlight button class when style equals highlight 1`] = `
<withHandlers(Component)
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  isOpen={false}
  style="highlight"
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(Component)>
`;

exports[`components/Dropdown should not render children when isDisabled is true 1`] = `
<withHandlers(Component)
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  idDisabled={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(Component)>
`;

exports[`components/Dropdown should not use .oui-arrow-inline--down when icon isEqual to triangle 1`] = `
<withHandlers(Component)
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(Component)>
`;

exports[`components/Dropdown should render children when isOpen is true 1`] = `
<withHandlers(Component)
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(Component)>
`;

exports[`components/Dropdown should use oui-button--full class when fullWidth is true 1`] = `
<withHandlers(Component)
  arrowIcon="none"
  buttonContent="Dropdown"
  fullWidth={true}
  icon={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(Component)>
`;
